Robust Dpatch

mathsinn opened the pull request:
# Description

Implementation of new class towards generating robust physical-world adversarial patches for object detectors.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Local proofs of concept (Jupyter notebook and example script)
- [x] Unit tests

**Test Configuration**:
- MacOS
- Python 3.7
- ART version 1.4.2
- PyTorch 1.6.0

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        ):
+            raise ValueError("The patch (partially) lies outside the cropped image.")
+
+        for i_step in trange(self.max_iter, desc="DPatchEOT iteration", disable=not self.verbose):
```
```suggestion
        for i_step in trange(self.max_iter, desc="RobustDPatch iteration", disable=not self.verbose):
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+
+        return self._patch
+
+    @deprecated_keyword_arg("channel_index", end_version="1.5.0", replaced_by="channels_first")
+    def _augment_images_with_patch(
+        self, x: np.ndarray, patch: np.ndarray, channels_first: bool, channel_index=Deprecated
+    ) -> Tuple[np.ndarray, np.ndarray, Dict[str, Union[int, float]]]:
```
```suggestion
    def _augment_images_with_patch(
        self, x: np.ndarray, patch: np.ndarray, channels_first: bool
    ) -> Tuple[np.ndarray, np.ndarray, Dict[str, Union[int, float]]]:
```

beat-buesser replied:
```suggestion
    ) -> Tuple[np.ndarray, List[Dict[str, np.ndarray]], Dict[str, Union[int, float]]]:
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        :param x: Sample images.
+        :param patch: The patch to be applied.
+        :param channels_first: Set channels first or last.
+        :param channel_index: Index of the color channel.
+        :type channel_index: `int`
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        :param channel_index: Index of the color channel.
+        :type channel_index: `int`
+        """
+        # Remove in 1.5.0
+        if channel_index == 3:
+            channels_first = False
+        elif channel_index == 1:
+            channels_first = True
+        elif channel_index is not Deprecated:
+            raise ValueError("Not a proper channel_index. Use channels_first.")
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+
+        return x_patch, patch_target, transformations
+
+    @deprecated_keyword_arg("channel_index", end_version="1.5.0", replaced_by="channels_first")
+    def _untransform_gradients(
+        self,
+        gradients: np.ndarray,
+        transforms: Dict[str, Union[int, float]],
+        channels_first: bool,
+        channel_index=Deprecated,
+    ) -> np.ndarray:
```
```suggestion
    def _untransform_gradients(
        self,
        gradients: np.ndarray,
        transforms: Dict[str, Union[int, float]],
        channels_first: bool,
    ) -> np.ndarray:
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        :param x: Sample images.
+        :param patch: The patch to be applied.
+        :param channels_first: Set channels first or last.
+        :param channel_index: Index of the color channel.
+        :type channel_index: `int`
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        :param channel_index: Index of the color channel.
+        :type channel_index: `int`
+        """
+        # Remove in 1.5.0
+        if channel_index == 3:
+            channels_first = False
+        elif channel_index == 1:
+            channels_first = True
+        elif channel_index is not Deprecated:
+            raise ValueError("Not a proper channel_index. Use channels_first.")
```
```suggestion
```

beat-buesser commented on tests/attacks/evasion/test_robust_dpatch.py:
```diff
@@ -0,0 +1,164 @@
+import numpy as np
+import pytest
+
+from art.attacks.evasion import RobustDPatch
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.object_detection.object_detector import ObjectDetectorMixin, ObjectDetector
+from tests.attacks.utils import backend_test_classifier_type_check_fail
+from tests.utils import master_seed
+from tests.utils import ARTTestException
```
```suggestion
from art.attacks.evasion import RobustDPatch
from art.estimators.estimator import BaseEstimator, LossGradientsMixin
from art.estimators.object_detection.object_detector import ObjectDetectorMixin, ObjectDetector

from tests.attacks.utils import backend_test_classifier_type_check_fail
from tests.utils import master_seed
from tests.utils import ARTTestException
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+from art.attacks.attack import EvasionAttack
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.object_detection.object_detector import ObjectDetectorMixin
+from art.utils import Deprecated, deprecated_keyword_arg
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        self.brightness_range = brightness_range
+        self.rotation_weights = rotation_weights
+        self.sample_size = sample_size
+        super()._check_params()
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        if x.shape[channel_index] != self.patch_shape[channel_index - 1]:
+            raise ValueError("The color channel index of the images and the patch have to be identical.")
+        if y is not None:
+            raise ValueError("The DPatchEOT attack does not use target labels.")
```
```suggestion
            raise ValueError("The RobustDPatch attack does not use target labels.")
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+            raise ValueError("The adversarial patch can only be applied to images.")
+
+        # Check whether patch fits into the cropped images:
+        if channel_index == 1:
+            image_height, image_width = x.shape[2:4]
+        else:
+            image_height, image_width = x.shape[1:3]
```
```suggestion
        if self.estimator.channels_first:
            image_height, image_width = x.shape[2:4]
        else:
            image_height, image_width = x.shape[1:3]
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        channel_index=Deprecated,
+    ) -> np.ndarray:
+        """
+        Augment images with patch.
+
+        :param x: Sample images.
+        :param patch: The patch to be applied.
```
```suggestion
        Revert transformation on gradients.

        :param gradients: The gradients to be reverse transformed.
        :param transforms: The transformations in forward direction.
```

beat-buesser commented on tests/attacks/evasion/test_robust_dpatch.py:
```diff
@@ -0,0 +1,164 @@
+# MIT License
```
Should we rename this file to `test_dpatch_robust.py` following the convention of prepending `test_` to the tested module's filename?

mathsinn replied:
I agree; I've renamed the file.

beat-buesser commented on tests/attacks/evasion/test_robust_dpatch.py:
```diff
@@ -0,0 +1,164 @@
+
+from art.attacks.evasion import RobustDPatch
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.object_detection.object_detector import ObjectDetectorMixin, ObjectDetector
```
```suggestion
from art.estimators.object_detection.object_detector import ObjectDetectorMixin
```

beat-buesser commented on tests/attacks/evasion/test_robust_dpatch.py:
```diff
@@ -0,0 +1,164 @@
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.estimators.object_detection.object_detector import ObjectDetectorMixin, ObjectDetector
+from tests.attacks.utils import backend_test_classifier_type_check_fail
+from tests.utils import master_seed
```
```suggestion
```

beat-buesser commented on art/attacks/evasion/dpatch_robust.py:
```diff
@@ -0,0 +1,432 @@
+        patch_location: Tuple[int, int] = (0, 0),
+        crop_range: Tuple[int, int] = (0, 0),
+        brightness_range: Tuple[float, float] = (1.0, 1.0),
+        rotation_weights: Union[Tuple[float], Tuple[int]] = (1, 0, 0, 0),
```
```suggestion
        rotation_weights: Union[Tuple[float, float, float, float], Tuple[int, int, int, int]] = (1, 0, 0, 0),
```

beat-buesser reviewed (changes requested):
Hi @mathsinn Thank you very much for implementing and testing a better, more robust version of  `DPatch`! I think it's a great name for the new attack. I have proposed a few updates above.

mathsinn commented:
Thank you @beat-buesser for the thorough review and the suggested changes. I've incorporated them all and also renamed the test module as per your suggestion.
